### PR TITLE
chore(profiling): update libdatadog to v10 and add crashtracker support

### PR DESCRIPTION
**Pull request by sanchda**
This PR adds crashtracker support and some tests.  This isn't yet able to be enabled at the product level.

I'll set up the support in a separate PR, this one was just getting way too ridiculously long.

No user-facing changes, so I'm thinking of not publishing a changelog.  This PR does not enable crashtracker.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/internal/datadog/profiling/cmake/FindCppcheck.cmake [hunk 1]
```diff
@@ -18,11 +18,6 @@ if (DO_CPPCHECK)
     )
     set(CPPCHECK_EXECUTABLE ${CMAKE_BINARY_DIR}/cppcheck/bin/cppcheck)
   endif()
-
-  # The function we use to register targets for cppcheck would require us to run separate
-  # commands for each target, which is annoying.  Instead we'll consolidate all the targets
-  # as dependencies of a single target, and then run that target.
-  add_custom_target(cppcheck_runner ALL COMMENT "Runs cppcheck on all defined targets")
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/cmake/FindCppcheck.cmake:25]: Why is this removed?

File: ddtrace/internal/datadog/profiling/crashtracker/_crashtracker.pyx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+@not_implemented
+def set_profiling_state_serializing(on: bool) -> None:
+    if on:
+        crashtracker_profiling_state_serializing_start()
+    else:
+        crashtracker_profiling_state_serializing_stop()
+
+
+@not_implemented
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/crashtracker/_crashtracker.pyx:158]: Should this be removed?

File: ddtrace/internal/datadog/profiling/util.py [hunk 3]
```diff
@@ -1,12 +1,23 @@
...
 
 LOG = get_logger(__name__)
 
 
+def ensure_binary_or_empty(s: StringType) -> bytes:
+    try:
+        return ensure_binary(s)
+    except Exception:
+        pass
+    return b""
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/util.py:18]: ci complaining

https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/64558/workflows/6c9df835-5137-48b5-823a-1cb055e4be10/jobs/3989355

>> Issue: [B110:try_except_pass] Try, Except, Pass detected.
   Severity: Low   Confidence: High
   CWE: CWE-703 (https://cwe.mitre.org/data/definitions/703.html)
   More Info: https://bandit.readthedocs.io/en/1.7.5/plugins/b110_try_except_pass.html
   Location: ddtrace/internal/datadog/profiling/util.py:16:4
15	        return ensure_binary(s)
16	    except Exception:
17	        pass
18	    return b""

File: ddtrace/internal/datadog/profiling/cmake/FindCppcheck.cmake (hunk 1 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/cmake/FindCppcheck.cmake:25]: Ah, because it's not used.  Later in the cmake you'll see the true target we actually use.  At one point I factored this differently and just didn't remove the original runner.
```
  # Create a standard target to run everything
  if (NOT TARGET cppcheck)
    add_custom_target(cppcheck COMMENT "Runs cppcheck on all projects")
  endif()
  if (DO_CPPCHECK)
    add_dependencies(cppcheck ${NAME})
  endif()
  ```

File: tests/internal/crashtracker/test_crashtracker.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        os.kill(os.getpid(), signal.SIGSEGV.value)
+        exit(-1)
+
+    # Part 5, check
+    conn = utils.listen_get_conn(sock)
+    assert conn
+
+    data = utils.conn_to_bytes(conn)
+    conn.close()
+    assert b"os_kill_impl" in data
```
Comment (taegyunkim) [inline tests/internal/crashtracker/test_crashtracker.py:232]: Where does the _impl come from?

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp [hunk 5]
```diff
@@ -136,8 +136,11 @@ Datadog::UploaderBuilder::build()
     }
 
     // If we're here, the tags are good, so we can initialize the exporter
-    ddog_prof_Exporter_NewResult res = ddog_prof_Exporter_new(
-      to_slice("dd-trace-py"), to_slice(profiler_version), to_slice(family), &tags, ddog_Endpoint_agent(to_slice(url)));
+    ddog_prof_Exporter_NewResult res = ddog_prof_Exporter_new(to_slice("dd-trace-py"),
+                                                              to_slice(profiler_version),
+                                                              to_slice(family),
+                                                              &tags,
+                                                              ddog_prof_Endpoint_agent(to_slice(url)));
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp:143]: format?

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/crashtracker_interface.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+crashtracker_set_receiver_binary_path(std::string_view path) // cppcheck-suppress unusedFunction
+{
+    return crashtracker.set_receiver_binary_path(path);
+}
+
+void
+crashtracker_start() // cppcheck-suppress unusedFunction
+{
+    // This is a one-time start pattern to ensure that the crashtracker is only started once.
+    const static bool initialized = []() {
```
Comment (taegyunkim) [inline ddtrace/internal/datadog/profiling/dd_wrapper/src/crashtracker_interface.cpp:116]: Don't we have to udpate `crashtracker_initialized` in this lambda or using this lambda?

File: tests/internal/crashtracker/test_crashtracker.py (hunk 4 above)
Comment (sanchda) [reply tests/internal/crashtracker/test_crashtracker.py:232]: It's an internal symbol in cpython.  I don't know precisely where it is defined, but I actually can't see it in the strings for cpython 3.12.  I'll drop the `_impl` part in this test.

File: ddtrace/internal/datadog/profiling/crashtracker/_crashtracker.pyx (hunk 2 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/crashtracker/_crashtracker.pyx:158]: I don't think so, but I do think the choice of the name for the decorator was quite poor.  I've factored this to be called `raise_if_unimplementable` in b559de2f9

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/crashtracker_interface.cpp (hunk 6 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/crashtracker_interface.cpp:116]: Ha!  I definitely missed this.  Fixed in eceb74d6f

File: ddtrace/internal/datadog/profiling/util.py (hunk 3 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/util.py:18]: Oops, I'm not running the security lints locally--thank you

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp (hunk 5 above)
Comment (sanchda) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp:143]: This is how our c++ style rules want to format the section, I think.  At least it appears to pass in CI.

If you're seeing something different on your local machine, it should be noted that clang-format linter appears to do  slightly different things depending on the version of clang.  I'm not 100% sure what CI uses.  Maybe 13?

**Review by taegyunkim — APPROVED** [review]
